src/client.ts

```diff
@@ -0,0 +1,222 @@
+   */
+  headers: Record<string, string>;
+}
+
+export interface MessageData<T = unknown> {
+  type: string;
+  requestId: string;
+  payload?: T;
+  error?: string;
+}
+
+type PendingRequest<T = unknown> = { resolve: (value: T) => void; reject: (reason: Error) => void };
+
+class Deferred<T = void> implements PromiseLike<T> {
+  promise: Promise<T>;
+  resolve: (value: T) => void = () => {};
+  reject: (reason?: Error) => void = () => {};
+
+  constructor() {
+    this.promise = new Promise<T>((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+  }
+
+  then<TResult1 = T, TResult2 = never>(
+    onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | null | undefined,
+    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null | undefined,
+  ): PromiseLike<TResult1 | TResult2> {
+    return this.promise.then(onfulfilled, onrejected);
+  }
+}
+
+export class ExtensionClient {
+  #requests = new Map<string, PendingRequest>();
+  #extensionIdPromise: Deferred<string | false> = new Deferred<string | false>();
+
+  constructor() {
+    this.sendMessage = this.sendMessage.bind(this);
+    window.addEventListener("message", this.#handleEvent, false);
+  }
+
+  #handleEvent = async (event: MessageEvent) => {
+    if (event.source !== window) {
+      return;
+    }
+
+    if (!event.data || typeof event.data !== "object") {
+      return;
+    }
+
+    const { requestId, type, error, payload } = event.data;
+
+    if (!type || typeof type !== "string") {
+      return;
+    }
+
+    if (type === "signify-extension") {
+      this.#extensionIdPromise.resolve(event.data.data.extensionId);
+      return;
+    }
+
+    if (type === "/signify/reply" && requestId && typeof requestId === "string") {
+      const promise = this.#requests.get(requestId);
+
+      if (!promise) {
+        return;
+      }
+
+      if (error) {
+        promise.reject(new Error(typeof error === "string" ? error : error.toString()));
+      } else if (!payload || typeof payload !== "object") {
+        promise.reject(new Error("No payload received in response"));
+      } else {
+        promise.resolve(payload);
+      }
+
+      this.#requests.delete(requestId);
+    }
+  };
+
+  isExtensionInstalled = async (timeout: number = 1000): Promise<string | false> => {
+    const timer = setTimeout(() => {
+      this.#extensionIdPromise.resolve(false);
+    }, timeout);
+
+    const result = await this.#extensionIdPromise;
+    clearTimeout(timer);
+
+    return result;
+  };
+
+  signRequest = async (req: SignRequestArgs): Promise<SignRequestResult> => {
+    return this.sendMessage("/signify/sign-request", req);
+  };
+
+  signData = async (payload: SignDataArgs): Promise<SignDataResult> => {
+    return this.sendMessage("/signify/sign-data", payload);
+  };
+
+  authorize = async (payload?: AuthorizeArgs): Promise<AuthorizeResult> => {
+    return this.sendMessage("/signify/authorize", payload);
```

Comment:
Currently in polaris-web, the api allows authorization with either credential (`requestCredential`: shows list of credentials to select one), aid (`requestAid`: shows list of aid to select one), or both (`requestAidORCred`: shows list of both creds and aids). We might need to add another method to distinguish this behavior (e.g `authorizeWithAid`).

Regarding `requestAutoSignin`:
There exists a feature named autoSign, the idea is to make a user select a sign in pair (combination or webapp origin + (aid OR credential)) with autoSignin flag as `true` (One origin can only have one signin pair with autoSignin true). So, the moment users request `requestAutoSignin`, they are not shown a list of credential/aid instead that particular signin pair gets selected. We'd also need to add this functionality.

Reply:
Is the `signData` intended to sign any arbitrary data? We should try to avoid requesting signature of any data for security concerns. It may be better to have specific method when requesting signatures so the extension can alert the user and request consent why such a signature is requested. Sign a request is one case of a specific signature. Others can be added, just as sign a document.

Reply:
Thanks for responding!

@rodo Yes. Currently the functionality is to sign arbitrary data. It is up to the extension implementation whether or not to ask user for consent before the signing is executed though. In our implementation the promise returned from signData from is resolved or rejected then based on user consent.

We also thought of having the user provide the data/documents in the extension, but the user journey was tricky to solve I think. It's up for grabs though... I agree with you rodo that we should avoid exposing a method to sign arbitrary data.

@hunni, thank you for clarifying! As you can see with a lot of the recent commits, we broke the compatibility. For example the autosignin. The plan is to finalise a minimum API surface that we are happy with, then try re-add compatibility again. Hopefully we are able to align.

In addition, we were thinking about requiring explicit user approval before injecting the content script. This would also make Google happier and probably speed up the review process because you don't need host permissions.